{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import numpy as np\n",
    "import neat\n",
    "import datetime\n",
    "import os\n",
    "import math\n",
    "\n",
    "from comunication_channel import AgentLogChannel\n",
    "\n",
    "from mlagents_envs.environment import UnityEnvironment\n",
    "from mlagents_envs.base_env import ActionTuple\n",
    "from mlagents_envs.side_channel.engine_configuration_channel import EngineConfigurationChannel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_node(g, config, new_node_id=None, bias=None, response=None, aggregation=None, activation=None):\n",
    "    if new_node_id is None:\n",
    "        new_node_id = config.get_new_node_key(g.nodes)\n",
    "    ng = g.create_node(config, new_node_id)\n",
    "    g.nodes[new_node_id] = ng\n",
    "    if bias is not None:\n",
    "        ng.bias = bias\n",
    "    if response is not None:\n",
    "        ng.response = response\n",
    "    if aggregation is not None:\n",
    "        ng.aggregation = aggregation\n",
    "    if activation is not None:\n",
    "        ng.activation = activation\n",
    "    return new_node_id, ng"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import neat\n",
    "from random import choice, gauss, random, uniform, randint\n",
    "from neat.genome import DefaultGenome\n",
    "from neat.genes import DefaultNodeGene, DefaultConnectionGene\n",
    "from neat.config import ConfigParameter, write_pretty_params\n",
    "from neat.genes import BaseGene\n",
    "from neat.attributes import FloatAttribute, BoolAttribute, StringAttribute, BaseAttribute\n",
    "from neat.activations import ActivationFunctionSet\n",
    "from neat.aggregations import AggregationFunctionSet\n",
    "from itertools import count\n",
    "\n",
    "class FixedFloatAttribute(BaseAttribute):\n",
    "    \"\"\"\n",
    "    Class for floating-point numeric attributes,\n",
    "    such as the response of a node or the weight of a connection.\n",
    "    \"\"\"\n",
    "    _config_items = {\"init_mean\": [float, None],\n",
    "                     \"init_stdev\": [float, None],\n",
    "                     \"init_type\": [str, 'gaussian'],\n",
    "                     \"replace_rate\": [float, None],\n",
    "                     \"mutate_rate\": [float, None],\n",
    "                     \"mutate_power\": [float, None],\n",
    "                     \"max_value\": [float, None],\n",
    "                     \"min_value\": [float, None]}\n",
    "\n",
    "    def clamp(self, value, config):\n",
    "        min_value = getattr(config, self.min_value_name)\n",
    "        max_value = getattr(config, self.max_value_name)\n",
    "        return max(min(value, max_value), min_value)\n",
    "\n",
    "    def init_value(self, config):\n",
    "        mean = getattr(config, self.init_mean_name)\n",
    "        stdev = getattr(config, self.init_stdev_name)\n",
    "        init_type = getattr(config, self.init_type_name).lower()\n",
    "\n",
    "        if ('gauss' in init_type) or ('normal' in init_type):\n",
    "            return self.clamp(gauss(mean, stdev), config)\n",
    "\n",
    "        if 'uniform' in init_type:\n",
    "            min_value = max(getattr(config, self.min_value_name),\n",
    "                            (mean - (2 * stdev)))\n",
    "            max_value = min(getattr(config, self.max_value_name),\n",
    "                            (mean + (2 * stdev)))\n",
    "            return uniform(min_value, max_value)\n",
    "\n",
    "        raise RuntimeError(f\"Unknown init_type {getattr(config, self.init_type_name)!r} for {self.init_type_name!s}\")\n",
    "\n",
    "    def mutate_value(self, value, config):\n",
    "        mutate_value = 1.0\n",
    "        # mutate_rate is usually no lower than replace_rate, and frequently higher -\n",
    "        # so put first for efficiency\n",
    "        mutate_rate = getattr(config, self.mutate_rate_name)\n",
    "\n",
    "        r = random()\n",
    "        if r < mutate_rate:\n",
    "            mutate_power = getattr(config, self.mutate_power_name)\n",
    "            return self.clamp(value + uniform(-mutate_value, mutate_value), config)\n",
    "\n",
    "        replace_rate = getattr(config, self.replace_rate_name)\n",
    "\n",
    "        r = random()\n",
    "        if r < replace_rate:\n",
    "            return self.init_value(config)\n",
    "\n",
    "        return value\n",
    "\n",
    "    def validate(self, config):\n",
    "        min_value = getattr(config, self.min_value_name)\n",
    "        max_value = getattr(config, self.max_value_name)\n",
    "        if max_value < min_value:\n",
    "            raise RuntimeError(\"Invalid min/max configuration for {self.name}\")\n",
    "\n",
    "class FixedStructureNodeGene(BaseGene):\n",
    "    _gene_attributes = [\n",
    "                        FloatAttribute('bias'),\n",
    "                        FloatAttribute('response'),\n",
    "                        StringAttribute('activation', options=''),\n",
    "                        StringAttribute('aggregation', options='')]\n",
    "\n",
    "    def distance(self, other, config):\n",
    "        return 0.0\n",
    "\n",
    "class FixedStructureConnectionGene(BaseGene):\n",
    "    _gene_attributes = [FixedFloatAttribute('weight'),\n",
    "                        BoolAttribute('enabled')]\n",
    "\n",
    "    def mutate(self, config):\n",
    "        # Mutate only weights\n",
    "        a = self._gene_attributes[0]\n",
    "        v = getattr(self, a.name)\n",
    "        setattr(self, a.name, a.mutate_value(v, config))\n",
    "\n",
    "    def distance(self, other, config):\n",
    "        d = abs(self.weight - other.weight)\n",
    "        return d\n",
    "    \n",
    "class FixedStructureGenomeConfig(object):\n",
    "    __params = [ConfigParameter('feed_forward', bool),\n",
    "                ConfigParameter('num_inputs', int),\n",
    "                ConfigParameter('num_outputs', int),\n",
    "                ConfigParameter('num_hidden', int),\n",
    "                ConfigParameter('num_hidden_layers', int),\n",
    "                ConfigParameter('initial_connection', str, 'full_direct'),\n",
    "                ConfigParameter('conn_add_prob', float),\n",
    "                ConfigParameter('conn_delete_prob', float),\n",
    "                ConfigParameter('node_add_prob', float),\n",
    "                ConfigParameter('node_delete_prob', float)]\n",
    "\n",
    "\n",
    "    def __init__(self, params):\n",
    "        # Create full set of available activation functions.\n",
    "        self.activation_defs = ActivationFunctionSet()\n",
    "        self.activation_options = params.get('activation_options', 'sigmoid').strip().split()\n",
    "        self.aggregation_options = params.get('aggregation_options', 'sum').strip().split()\n",
    "        self.aggregation_function_defs = AggregationFunctionSet()\n",
    "\n",
    "        # Gather configuration data from the gene classes.\n",
    "        self.node_gene_type = params['node_gene_type']\n",
    "        self.__params += self.node_gene_type.get_config_params()\n",
    "        self.connection_gene_type = params['connection_gene_type']\n",
    "        self.__params += self.connection_gene_type.get_config_params()\n",
    "        self.connection_fraction = None\n",
    "        self.node_indexer = None\n",
    "        self.structural_mutation_surer = 'false'\n",
    "\n",
    "        # Use the configuration data to interpret the supplied parameters.\n",
    "        for p in self.__params:\n",
    "            setattr(self, p.name, p.interpret(params))\n",
    "\n",
    "        # By convention, input pins have negative keys, and the output\n",
    "        # pins have keys 0,1,...\n",
    "        self.input_keys = [-i - 1 for i in range(self.num_inputs)]\n",
    "        self.output_keys = [i for i in range(self.num_outputs)]\n",
    "\n",
    "    def get_new_node_key(self, node_dict):\n",
    "        if self.node_indexer is None:\n",
    "            if node_dict:\n",
    "                self.node_indexer = count(max(list(node_dict)) + 1)\n",
    "            else:\n",
    "                self.node_indexer = count(max(list(node_dict)) + 1)\n",
    "\n",
    "        new_id = next(self.node_indexer)\n",
    "\n",
    "        assert new_id not in node_dict\n",
    "\n",
    "        return new_id\n",
    "    \n",
    "    def save(self, f):\n",
    "        write_pretty_params(f, self, self.__params)\n",
    "\n",
    "class FixedStructureGenome(object):\n",
    "    @classmethod\n",
    "    def parse_config(cls, param_dict):\n",
    "        param_dict['node_gene_type'] = FixedStructureNodeGene\n",
    "        param_dict['connection_gene_type'] = FixedStructureConnectionGene\n",
    "        return FixedStructureGenomeConfig(param_dict)\n",
    "    \n",
    "    @classmethod\n",
    "    def write_config(cls, f, config):\n",
    "        config.save(f)\n",
    "\n",
    "    def __init__(self, key):\n",
    "        self.key = key\n",
    "        self.connections = {}\n",
    "        self.nodes = {}\n",
    "\n",
    "        # Fitness results.\n",
    "        self.fitness = None\n",
    "\n",
    "    def mutate(self, config):\n",
    "        \"\"\" Mutates this genome. \"\"\"\n",
    "        # Mutate connection genes.\n",
    "        for cg in self.connections.values():\n",
    "            cg.mutate(config)\n",
    "\n",
    "        # Mutate node genes (bias, response, etc.).\n",
    "        for ng in self.nodes.values():\n",
    "            ng.mutate(config)\n",
    "\n",
    "    def configure_crossover(self, genome1, genome2, config):\n",
    "        \"\"\" Configure a new genome by crossover from two parent genomes. \"\"\"\n",
    "        if genome1.fitness > genome2.fitness:\n",
    "            parent1, parent2 = genome1, genome2\n",
    "        else:\n",
    "            parent1, parent2 = genome2, genome1\n",
    "\n",
    "        # Inherit connection genes\n",
    "        for key, cg1 in parent1.connections.items():\n",
    "            cg2 = parent2.connections.get(key)\n",
    "            if cg2 is None:\n",
    "                # Excess or disjoint gene: copy from the fittest parent.\n",
    "                self.connections[key] = cg1.copy()\n",
    "            else:\n",
    "                # Homologous gene: combine genes from both parents.\n",
    "                self.connections[key] = cg1.crossover(cg2)\n",
    "            self.connections[key].enabled = True\n",
    "\n",
    "        # Inherit node genes\n",
    "        parent1_set = parent1.nodes\n",
    "        parent2_set = parent2.nodes\n",
    "\n",
    "        for key, ng1 in parent1_set.items():\n",
    "            ng2 = parent2_set.get(key)\n",
    "            assert key not in self.nodes\n",
    "            if ng2 is None:\n",
    "                # Extra gene: copy from the fittest parent\n",
    "                self.nodes[key] = ng1.copy()\n",
    "            else:\n",
    "                # Homologous gene: combine genes from both parents.\n",
    "                self.nodes[key] = ng1.crossover(ng2)\n",
    "\n",
    "    # def add_connection(self, config, input_key, output_key, weight):\n",
    "    #     key = (input_key, output_key)\n",
    "    #     connection = FixedStructureConnectionGene(key)\n",
    "    #     connection.init_attributes(config)\n",
    "    #     connection.weight = weight\n",
    "    #     connection.enabled = True\n",
    "\n",
    "    #     self.connections[key] = connection\n",
    "\n",
    "    def distance(self, other, config):\n",
    "        return 0.0\n",
    "    \n",
    "    def size(self):\n",
    "        \"\"\"Returns genome 'complexity', taken to be (number of nodes, number of enabled connections)\"\"\"\n",
    "        num_enabled_connections = sum([1 for cg in self.connections.values() if cg.enabled is True])\n",
    "        return len(self.nodes), num_enabled_connections\n",
    "\n",
    "    def __str__(self):\n",
    "        s = f\"Key: {self.key}\\nFitness: {self.fitness}\\nNodes:\"\n",
    "        for k, ng in self.nodes.items():\n",
    "            s += f\"\\n\\t{k} {ng!s}\"\n",
    "        s += \"\\nConnections:\"\n",
    "        connections = list(self.connections.values())\n",
    "        connections.sort()\n",
    "        for c in connections:\n",
    "            s += \"\\n\\t\" + str(c)\n",
    "        return s\n",
    "    \n",
    "    # def add_hidden_nodes(self, config):\n",
    "    #     for i in range(config.num_hidden):\n",
    "    #         node_key = self.get_new_node_key()\n",
    "    #         assert node_key not in self.nodes\n",
    "    #         node = self.__class__.create_node(config, node_key)\n",
    "    #         self.nodes[node_key] = node\n",
    "\n",
    "    def configure_new(self, config):\n",
    "        # Create node genes for the input and output nodes\n",
    "        for node_key in config.input_keys + config.output_keys:\n",
    "            self.nodes[node_key] = self.create_node(config, node_key)\n",
    "\n",
    "        # Initialize list to keep track of nodes in the last layer (start with input layer)\n",
    "        last_layer_nodes = config.input_keys\n",
    "\n",
    "        # Add hidden nodes and layers\n",
    "        # config.num_hidden now represents the number of neurons PER HIDDEN LAYER\n",
    "        if config.num_hidden > 0 and config.num_hidden_layers > 0:\n",
    "            for _ in range(config.num_hidden_layers):\n",
    "                layer_node_ids = []\n",
    "                for _ in range(config.num_hidden):  # Create the specified number of nodes for this layer\n",
    "                    node_key = config.get_new_node_key(self.nodes)\n",
    "                    self.nodes[node_key] = self.create_node(config, node_key)\n",
    "                    layer_node_ids.append(node_key)\n",
    "                \n",
    "                # Connect all nodes from the last layer to the current layer's nodes\n",
    "                for last_layer_node in last_layer_nodes:\n",
    "                    for current_layer_node in layer_node_ids:\n",
    "                        connection = self.create_connection(config, last_layer_node, current_layer_node)\n",
    "                        self.connections[connection.key] = connection\n",
    "\n",
    "                # Update last_layer_nodes to reflect the current layer for next iteration\n",
    "                last_layer_nodes = layer_node_ids\n",
    "\n",
    "            # Connect all nodes from the last hidden layer to the output nodes\n",
    "            for hidden_node_id in last_layer_nodes:\n",
    "                for output_node_id in config.output_keys:\n",
    "                    connection = self.create_connection(config, hidden_node_id, output_node_id)\n",
    "                    self.connections[connection.key] = connection\n",
    "        else:\n",
    "            # If there are no hidden layers, connect input nodes directly to output nodes\n",
    "            for input_node_id in config.input_keys:\n",
    "                for output_node_id in config.output_keys:\n",
    "                    connection = self.create_connection(config, input_node_id, output_node_id)\n",
    "                    self.connections[connection.key] = connection\n",
    "            \n",
    "    @staticmethod\n",
    "    def create_node(config, node_id):\n",
    "        node = FixedStructureNodeGene(node_id)\n",
    "        node.init_attributes(config)\n",
    "        return node\n",
    "\n",
    "    @staticmethod\n",
    "    def create_connection(config, input_id, output_id):\n",
    "        connection = config.connection_gene_type((input_id, output_id))\n",
    "        connection.init_attributes(config)\n",
    "        return connection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "NUM_RUNS = 1\n",
    "MAX_GENS = 1000\n",
    "generation = 0\n",
    "SAVE_INTERVAL = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "engine_config_channel = EngineConfigurationChannel()\n",
    "engine_config_channel.set_configuration_parameters(time_scale=10)\n",
    "agent_count_channel = AgentLogChannel()\n",
    "\n",
    "env_path = \"../Builds/train-env/autonomous-drone.exe\"\n",
    "save_nn_destination = 'result/best.pkl'\n",
    "\n",
    "env = UnityEnvironment(file_name=None, worker_id=0, no_graphics=True, side_channels=[engine_config_channel, agent_count_channel])\n",
    "env.reset()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Behaviour specs <mlagents_envs.base_env.BehaviorMapping object at 0x000001DE7537F9D0>\n",
      "Name of the behavior : DroneControl?team=0\n",
      "Number of observations :  1\n",
      "There are 4 continuous actions\n"
     ]
    }
   ],
   "source": [
    "out_mult = 1\n",
    "\n",
    "behavior_specs = env.behavior_specs\n",
    "print(f\"Behaviour specs {behavior_specs}\")\n",
    "behavior_name = list(behavior_specs)[0]\n",
    "spec = env.behavior_specs[behavior_name]\n",
    "\n",
    "print(f\"Name of the behavior : {behavior_name}\")\n",
    "print(\"Number of observations : \", len(spec.observation_specs)) # vector if 1\n",
    "\n",
    "# Is the Action continuous or multi-discrete ?\n",
    "if spec.action_spec.continuous_size > 0:\n",
    "  print(f\"There are {spec.action_spec.continuous_size} continuous actions\")\n",
    "if spec.action_spec.is_discrete():\n",
    "  print(f\"There are {spec.action_spec.discrete_size} discrete actions\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_folder(file_name_prefix):\n",
    "    directory = os.path.dirname(f\"{file_name_prefix}\")\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_agents_and_double_reset(num_agents: int):\n",
    "    agent_count_channel.send_int(data=num_agents) \n",
    "    env.reset()\n",
    "    env.reset()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_observation_for_agent(agent: int, observations):\n",
    "    for observation in observations:\n",
    "        key = int(observation.split(\"=\")[2])\n",
    "        if key == agent:\n",
    "            return observations[observation]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_agent_ids(decision_steps):\n",
    "    \"\"\"\n",
    "    Map agent ids between NEAT and UNITY.\n",
    "\n",
    "    Args:\n",
    "        decision_steps: An iterable containing decision steps.\n",
    "\n",
    "    Returns:\n",
    "        A tuple of two dictionaries: (unity_to_neat_map, neat_to_unity_map)\n",
    "    \"\"\"\n",
    "    unity_to_neat_map = {}\n",
    "    neat_to_unity_map = {}\n",
    "    id_count = 0\n",
    "    for step in decision_steps:\n",
    "        unity_to_neat_map[step] = id_count\n",
    "        neat_to_unity_map[id_count] = step\n",
    "        id_count += 1\n",
    "    return unity_to_neat_map, neat_to_unity_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def create_policies(genomes, cfg):\n",
    "    policies = []\n",
    "    for _, g in genomes:\n",
    "        g.fitness = 0\n",
    "        policy = neat.nn.FeedForwardNetwork.create(g, cfg)\n",
    "        policies.append(policy)\n",
    "    return policies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def eval_genomes(genomes, cfg):\n",
    "    global generation \n",
    "    generation += 1\n",
    "    policies = create_policies(genomes, cfg)\n",
    "    set_agents_and_double_reset(len(policies))\n",
    "    decision_steps, terminal_steps = env.get_steps(behavior_name)\n",
    "    agent_count = len(decision_steps.agent_id)\n",
    "\n",
    "    unity_to_neat_map, neat_to_unity_map = map_agent_ids(decision_steps)\n",
    "\n",
    "    done = False  # Vectorized initialization\n",
    "    removed_agents = []\n",
    "\n",
    "    episode_rewards = [0] * agent_count\n",
    "    print(f\"Agent count: {agent_count}\")\n",
    "\n",
    "    while not done:\n",
    "        for agent in decision_steps:\n",
    "            if unity_to_neat_map[agent] not in removed_agents:\n",
    "                nn_input =  np.asarray(decision_steps[agent].obs[:])\n",
    "                #print(f\"NN INPUT: {nn_input}, agent: {agent}\")\n",
    "                actions = policies[unity_to_neat_map[agent]].activate(nn_input[0])\n",
    "                continous_actions = np.asarray([actions])\n",
    "                continous_actions = np.clip(continous_actions, -1, 1)\n",
    "                #print(continous_actions)\n",
    "                action_tuple = ActionTuple(discrete=None, continuous=continous_actions)\n",
    "                env.set_action_for_agent(behavior_name=behavior_name, \n",
    "                                        agent_id=agent, \n",
    "                                        action=action_tuple)\n",
    "        env.step()\n",
    "        decision_steps, terminal_steps = env.get_steps(behavior_name)\n",
    "        for agent in range(agent_count):\n",
    "            if agent not in removed_agents:\n",
    "                local_agent = neat_to_unity_map[agent]\n",
    "                if local_agent in terminal_steps:\n",
    "                    episode_rewards[agent] += terminal_steps[local_agent].reward\n",
    "                    removed_agents.append(agent)\n",
    "                    #print(f\"Finished: {agent}\")\n",
    "                elif local_agent in decision_steps:\n",
    "                    episode_rewards[agent] += decision_steps[local_agent].reward\n",
    "                 \n",
    "\n",
    "        if len(removed_agents) >= agent_count:\n",
    "            print(\".\") \n",
    "            done = True\n",
    "    for i, (_, genome) in enumerate(genomes):\n",
    "        genome.fitness = episode_rewards[i]\n",
    "    if generation % SAVE_INTERVAL == 0:\n",
    "      global file_name_prefix\n",
    "      best_genome_current_generation = max(genomes, key=lambda x: x[1].fitness) \n",
    "      with open(file_name_prefix+'best-'+str(generation)+'.pkl', 'wb') as f:\n",
    "          pickle.dump(best_genome_current_generation, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(config_file, run, datte):\n",
    "    print(f\"Running {run}\")\n",
    "    global file_name_prefix\n",
    "    file_name_prefix = f\"checkpoints/{datte}/run-{run}/\"\n",
    "    create_folder(file_name_prefix=file_name_prefix)\n",
    "\n",
    "    config = neat.Config(FixedStructureGenome, neat.DefaultReproduction,\n",
    "                         neat.DefaultSpeciesSet, neat.DefaultStagnation,\n",
    "                         config_file)\n",
    "    \n",
    "    pop = neat.Population(config)\n",
    "\n",
    "    stats = neat.StatisticsReporter()\n",
    "\n",
    "    pop.add_reporter(stats)\n",
    "    #pop.add_reporter(neat.Checkpointer(generation_interval=5, time_interval_seconds=100000, filename_prefix=file_name_prefix))\n",
    "    pop.add_reporter(neat.TBReporter(True, 0, run, datte))\n",
    "    #pop.add_reporter(neat.StdOutReporter(True))\n",
    "    best = pop.run(eval_genomes, MAX_GENS)\n",
    "    # Display the winning genome.\n",
    "    print('\\nBest genome:\\n{!s}'.format(best))\n",
    "    print(\"Finished running!\")\n",
    "    \n",
    "    # Save best genome\n",
    "    with open(f'logs/{datte}/{run}/best.pkl', 'wb') as f:\n",
    "        pickle.dump(best, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ****** Running generation 0 ****** \n",
      "\n",
      "Agent count: 50\n",
      ".\n",
      "Population's average fitness: 679.94783 stdev: 0.00000\n",
      "Best fitness: 679.94783 - size: (0, 0) - species 1 - id 1\n",
      "Best fitness ever found! : 679.9478325843811\n",
      "Population of 50 members in 1 species:\n",
      "   ID   age  size   fitness   adj fit  stag\n",
      "  ====  ===  ====  =========  =======  ====\n",
      "     1    0    50  679.9478325843811      0.0     0\n",
      "Total extinctions: 0\n",
      "Generation time: 7.483 sec\n",
      "\n",
      " ****** Running generation 1 ****** \n",
      "\n",
      "Agent count: 50\n",
      ".\n",
      "Population's average fitness: 681.45670 stdev: 0.00000\n",
      "Best fitness: 681.45670 - size: (0, 0) - species 1 - id 1\n",
      "Best fitness ever found! : 681.4567031860352\n",
      "Population of 50 members in 1 species:\n",
      "   ID   age  size   fitness   adj fit  stag\n",
      "  ====  ===  ====  =========  =======  ====\n",
      "     1    1    50  681.4567031860352      0.0     0\n",
      "Total extinctions: 0\n",
      "Generation time: 4.242 sec (5.863 average)\n",
      "\n",
      " ****** Running generation 2 ****** \n",
      "\n",
      "Agent count: 50\n"
     ]
    }
   ],
   "source": [
    "config_path = 'config_fixed'\n",
    "datte = datetime.datetime.now().strftime(\"%d-%m-%Y--%H_%M\")\n",
    "# pre create folders for checkpointer\n",
    "for r in range(NUM_RUNS):\n",
    "    run(config_path, r, datte)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "env.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Best genome:\n",
      "(42511, <neat.genome.DefaultGenome object at 0x0000023FAF3574C0>)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'tuple' object has no attribute 'fitness'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[16], line 5\u001b[0m\n\u001b[0;32m      3\u001b[0m     genome \u001b[38;5;241m=\u001b[39m pickle\u001b[38;5;241m.\u001b[39mload(f)\n\u001b[0;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mBest genome:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;132;01m{!s}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(genome))\n\u001b[1;32m----> 5\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mgenome\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfitness\u001b[49m)\n\u001b[0;32m      6\u001b[0m config \u001b[38;5;241m=\u001b[39m neat\u001b[38;5;241m.\u001b[39mConfig(FixedStructureGenome, neat\u001b[38;5;241m.\u001b[39mDefaultReproduction,\n\u001b[0;32m      7\u001b[0m                     neat\u001b[38;5;241m.\u001b[39mDefaultSpeciesSet, neat\u001b[38;5;241m.\u001b[39mDefaultStagnation,\n\u001b[0;32m      8\u001b[0m                     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfixed_structure_config\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m      9\u001b[0m visualize\u001b[38;5;241m.\u001b[39mdraw_net(config, genome, \u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'tuple' object has no attribute 'fitness'"
     ]
    }
   ],
   "source": [
    "import visualize\n",
    "with open(save_nn_destination, \"rb\") as f:\n",
    "    genome = pickle.load(f)\n",
    "    print('\\nBest genome:\\n{!s}'.format(genome))\n",
    "print(genome.fitness)\n",
    "config = neat.Config(FixedStructureGenome, neat.DefaultReproduction,\n",
    "                    neat.DefaultSpeciesSet, neat.DefaultStagnation,\n",
    "                    'fixed_config')\n",
    "visualize.draw_net(config, genome, True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mlagents",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
